### PR TITLE
Jr invoicing

### PR DESCRIPTION
**Pull request by jjrr1717**
After Paypal payment the following occurs:
1. decrease inventory in Lot tables in database
2. apply discount in Account_Discount table in database
3. save invoice in Invoice table in database 

Removed tokens. 
Checks if info from previous step is present. 
If not, sends 406 to front end. 
If yes, moves forward. 

Checkout flow is complete.

### CODE REVIEW

**Comment by matthewjflee** [comment]
Json created at view cart:
https://user-images.githubusercontent.com/49124705/76172262-89cb3100-6159-11ea-98da-10b7d4f913b1.png

Json created at review order:
https://user-images.githubusercontent.com/49124705/76172252-7ae47e80-6159-11ea-83d8-2d4c461cc35a.png

Test that the tokens work as well.
-Can't go to shipping order unless added and viewed cart.
-Can't go to apply discount without viewing and entering shipping information
-Can go to payment page without viewing discount page

Test that can't over order product.

Thank you!

**Comment by jjrr1717** [comment]
Some paypal Information: 

URL to make payment:
@PostMapping("/customer/cart/payment")

URL's for successful or unsuccessful payments:
public static final String SUCCESS_URL = "cart/payment/success";
public static final String CANCEL_URL = "cart/payment";

File: src/main/java/ca/sait/vezorla/controller/PaypalController.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+@CrossOrigin
+public class PaypalController {
+
+    public static final String SUCCESS_URL = "cart/payment/success";
+    public static final String CANCEL_URL = "cart/payment";
+    PaypalServices paypalServices;
+
+    private UserServices userServices;
+
+    public PaypalController(PaypalServices paypalServices, UserServices userServices){
```
Comment (matthewjflee) [inline src/main/java/ca/sait/vezorla/controller/PaypalController.java]: Please add `@AllArgsConstructor` and remove the constructor

File: src/main/java/ca/sait/vezorla/controller/PaypalController.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+        } catch (PayPalRESTException e) {
+            e.printStackTrace();
+        }
+        System.out.println("Made payment");
+
+        return "redirect:/";
+    }
+
+    @GetMapping(value = "/customer/cart/payment/cancel")
```
Comment (matthewjflee) [inline src/main/java/ca/sait/vezorla/controller/PaypalController.java]: `@PostMapping` ? I think

File: src/main/java/ca/sait/vezorla/repository/AccountDiscountRepo.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+public class AccountDiscountRepo {
+
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    @Transactional
+    public void insertWithQuery(AccountDiscount accountDiscount) {
+        entityManager.createNativeQuery("INSERT INTO account_discount (account_email, discount_code) VALUES (?,?)")
+                .setParameter(1, accountDiscount.getEmail().getEmail())
+                .setParameter(2, accountDiscount.getCode().getCode())
```
Comment (matthewjflee) [inline src/main/java/ca/sait/vezorla/repository/AccountDiscountRepo.java]: We did a good job here 😊

File: src/main/java/ca/sait/vezorla/service/PaypalServices.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ *
+ * Taken from: https://github.com/Java-Techie-jt/spring-boot-paypal-example/blob/master/src/main/java/com/javatechie/spring/paypal/api/PaypalService.java
+ */
+@Service
+public class PaypalServices {
+
+    private APIContext apiContext;
+
+    public PaypalServices(APIContext apiContext){
+        this.apiContext = apiContext;
```
Comment (matthewjflee) [inline src/main/java/ca/sait/vezorla/service/PaypalServices.java]: Can this be `@AllArgsConstructor`? I tried this with `CustomerClientUtil` and it didn't work

File: src/main/java/ca/sait/vezorla/controller/PaypalController.java (hunk 1 above)
Comment (jjrr1717) [reply src/main/java/ca/sait/vezorla/controller/PaypalController.java]: Done

File: src/main/java/ca/sait/vezorla/repository/AccountDiscountRepo.java (hunk 3 above)
Comment (jjrr1717) [reply src/main/java/ca/sait/vezorla/repository/AccountDiscountRepo.java]: Yes we did!!

File: src/main/java/ca/sait/vezorla/controller/PaypalController.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.*;
+
+import javax.persistence.EntityManager;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpSession;
+
+/**
+ * Controller for paypal payments
+ * JR is the number one
```
Comment (matthewjflee) [inline src/main/java/ca/sait/vezorla/controller/PaypalController.java]: Wow!!!!

File: src/main/java/ca/sait/vezorla/service/PaypalServices.java (hunk 4 above)
Comment (jjrr1717) [reply src/main/java/ca/sait/vezorla/service/PaypalServices.java]: Done

File: src/main/java/ca/sait/vezorla/controller/PaypalController.java (hunk 2 above)
Comment (jjrr1717) [reply src/main/java/ca/sait/vezorla/controller/PaypalController.java]: I don't think so.  We are not doing anything with any data.  
Are you sure?

Comment (matthewjflee) [reply src/main/java/ca/sait/vezorla/controller/PaypalController.java]: I have no idea. Let's keep it

File: src/main/java/ca/sait/vezorla/controller/PaypalController.java (hunk 5 above)
Comment (jjrr1717) [reply src/main/java/ca/sait/vezorla/controller/PaypalController.java]: Honestly - it came from 
https://github.com/Java-Techie-jt/spring-boot-paypal-example/blob/master/src/main/java/com/javatechie/spring/paypal/api/PaypalController.java
So, I can't take credit.

**Comment by matthewjflee** [comment]
JR cannot commit changes anymore. Closing PR

**Pull request closed without merging**